### PR TITLE
glue must be compiled for the host, not the target.

### PR DESCRIPTION
**Pull request by SirVer**
On master, glue is compiled for the target platform. But since it generates `glue.rs`, it needs to be executable on the host platform instead. 

This patch changes `build.rs`, so that glue is build for the `$HOST` platform instead of the `$TARGET` platform.

### CODE REVIEW

**Comment by SirVer** [comment]
Checking in various glues works for me. Here is the commit implementing that: https://github.com/SirVer/rust-lua53/commit/71b9284efbced9c0c24cc105f0317fd3a0cb51b3

And here is a Travis run cross compiling a crate using `rust-lua53` for arm, linux x86 and linux x86_64 and mac os: https://travis-ci.org/SirVer/shell_grunt2/builds/331304618

So this is a bit ugly - I created these files by compiling this crate on the host machines directly. This generated the correct `glue.rs` which I then copied and checked into my repository. This means that this crate can not be compiled anymore on crates for which no `glue.rs` is checked in, but it also enables cross compilation which was impossible before.

@blt Your call if this is something you'd want in mond. It is quite a tradeoff.

**Comment by blt** [comment]
@SirVer I'm for it. Seems reasonably straightforward to add new targets. If you'd be so kind as to open a PR on mond I'd merge without much concern, maybe with instructions in the README so people can generate new targets if needed?

**Comment by jcmoyer** [comment]
Is it possible to cross-compile glue.c for $TARGET using gcc? That way we would be able to avoid checking in autogenerated files.

**Comment by SirVer** [comment]
@blt Feel free to use this patch for mond. I'd prefer to upstream it to this repo, since it is canonical and widely used.

@jcmoyer It is possible to cross compile - and in fact, we do this right now, i.e. before this PR. Unfortunately, it won't help since the cross compiled binary does not run on $HOST.

I think there is really no way around defining these constants for various platforms. However, I also think that `glue.rs` is probably not really necessary to be generated at all. For example, the diff between [`glue_darwin_86_64.rs`](https://github.com/SirVer/rust-lua53/blob/71b9284efbced9c0c24cc105f0317fd3a0cb51b3/src/ffi/glue_darwin_x86_64.rs) (64 bit, intel, Mac OS) and [`glue_linux_arm_gnu.rs`](https://github.com/SirVer/rust-lua53/blob/71b9284efbced9c0c24cc105f0317fd3a0cb51b3/src/ffi/glue_linux_arm_gnu.rs) (32bit, arm, linux) which are architecturally pretty much as far apart as possible is tiny. The only two differences are `LUA_EXTRASPACE` (8 vs 4) and `LUA_KCONTEXT` (`intptr_t` vs `ptrdiff_t`) - everything else is identical. So maybe we can check in a standard `glue.rs` and only overwrite these two types for platforms?

**Comment by jcmoyer** [comment]
The only problem I can think of with ditching the codegen approach (aside from what has already been mentioned) is that sometimes people modify luaconf.h to fit their needs (e.g. to disable floating point numbers entirely). I'm not sure how well that integrates with the rust ecosystem though, or if it's even worth supporting.

**Comment by SirVer** [comment]
@jcmoyer is that not the same as before? I.e. you have to modify the generated file or header in place to fit your needs. Now you would need to modify the checked in files.

**Comment by SirVer** [comment]
Seems like there is not much more movement on this. @jcmoyer Could you make a call? Do you want to merge this or not? I am not offended if this is closed as out of scope :)

**Comment by jcmoyer** [comment]
Sorry for the late response @SirVer. I'd like to elaborate a bit more on my stance.

Right now if you want to compile against a reconfigured Lua implementation, all you have to do is point `$LUA_LOCAL_SOURCE` to the appropriate directory and the code generator will leverage the C preprocessor to pick up all the values in `luaconf.h` without issue. The reason I originally took this approach is because Lua has a bunch of typedefs and logic implemented in or behind C macros, which unfortunately do not play nicely in the Rust ecosystem.

I do not like the proposed solution for two reasons:

Firstly, if upstream Lua makes changes to `luaconf.h` (e.g. changing defaults or adding more macro branches) we'd have to make sure all of the `glue_$PLATFORM.rs` files are in sync, which means re-running the glue generator on each platform and checking in the result on each. This assumes that we still want to leverage the code generator and C preprocessor to reduce the potential for human error. This may be automatable, but I do not know how to go about it.

Secondly, modifying the generated code by hand defeats the point of generating it in the first place:

* It's error prone. For example, upstream `luaconf.h` could be changed and the changes not caught by us or a user of a reconfigured `glue.rs`.
* If someone wants to ship on multiple platforms he'll have to make changes to multiple `glue_$PLATFORM.rs` files whereas currently he only needs to modify one file: `luaconf.h`.

In summary, it comes down to lack of maintainability. I'm open to reconsidering if it can be shown that we don't actually need to do all sorts of manual housekeeping as Lua is updated or more platforms are supported.

I've given the problem some thought since it was brought up, and one possible (but crazy) way to solve this would be to invoke `$TARGET-gcc`'s preprocessor for `luaconf.h` to eliminate the conditional macros. Its output could then be processed on `$HOST` as a codegen step. This would eliminate the need to run a `$TARGET` code generator, but it would not be a trivial task.

**Comment by SirVer** [comment]
Fair enough - your points make sense to me. I think your last paragraph makes the most sense for properly fixing the cross compilation issue, but is nothing I can invest time into right now. I will keep this change in my fork and depend on my fork for the while being in my projects that require cross compilation.

Thanks!
